File: module/backend-system/src/main/java/io/github/faustofanb/admin/module/system/domain/entity/SysRole.java
package io.github.faustofanb.admin.module.system.domain.entity;

import io.github.faustofanb.admin.core.domain.BaseEntity;
import org.babyfish.jimmer.sql.Entity;
import org.babyfish.jimmer.sql.ManyToMany;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * 角色表
 */
@Entity
public interface SysRole extends BaseEntity {

    /**
     * 角色名称
     */
    String roleName();

    /**
     * 角色权限字符串
     */
    String roleKey();

    /**
     * 显示顺序
     */
    int roleSort();

    /**
     * 数据范围（1：全部数据权限 2：自定数据权限 3：本部门数据权限 4：本部门及以下数据权限）
     */
    @Nullable
    String dataScope();

    /**
     * 角色状态（0正常 1停用）
     */
    int status();

    /**
     * 备注
     */
    @Nullable
    String remark();

    /**
     * 菜单组
     */
    @ManyToMany
    List<SysMenu> menus();

    /**
     * 部门组（数据权限）
     */
    @ManyToMany
    List<SysDept> depts();
}

File: src/main/kotlin/io/github/faustofan/admin/system/domain/entity/SysRole.kt
package io.github.faustofan.admin.system.domain.entity

import io.github.faustofan.admin.common.context.DataScope
import io.github.faustofan.admin.common.model.TenantAware
import org.babyfish.jimmer.sql.*

@Entity
interface SysRole : TenantAware {

    @Key
    val code: String

    val name: String

    val description: String

    // 1. 关联菜单 (控制页面可见性)
    @ManyToMany
    @JoinTable(
        name = "sys_role_menu_mapping",
        joinColumnName = "role_id",
        inverseJoinColumnName = "menu_id"
    )
    val menus: List<SysMenu>

    // 2. 关联权限 (控制API调用能力)
    // 中间表自动生成: sys_role_permission_mapping
    @ManyToMany
    @JoinTable(
        name = "sys_role_permission_mapping",
        joinColumnName = "role_id",
        inverseJoinColumnName = "permission_id"
    )
    val permissions: List<SysPermission>

    @ManyToMany(mappedBy = "roles")
    val users: List<SysUser>

    // 数据范围策略
    val dataScope: DataScope

    // 当 dataScope = CUSTOM 时，关联具体的部门
    @ManyToMany
    @JoinTable(name = "sys_role_org_mapping")
    val specificOrgs: List<SysOrg>
}

